Userland/Libraries/LibBitTorrent/Bencode/BDecoder.cpp

```diff
@@ -0,0 +1,105 @@
+    }
+
+    return Error::from_string_literal("Can't parse type");
+}
+
+ErrorOr<i64> BDecoder::parse_integer(Stream& stream)
+{
+    auto integer_str = StringBuilder();
+    u8 digit;
+    while ((digit = TRY(stream.read_value<u8>())) != 'e') {
+        if (is_digit(digit) || digit == '-') {
+            if (digit == '-' && integer_str.length() != 0)
+                return Error::from_string_literal("Invalid integer: When used, minus sign must be the first character.");
+            else if (integer_str.string_view() == "0")
+                return Error::from_string_literal("Invalid integer: Leading 0s not allowed.");
+            else if (integer_str.string_view() == '-' && digit == '0')
+                return Error::from_string_literal("Invalid integer: Leading 0s and -0 not allowed.");
+
+            integer_str.append(digit);
+        } else {
+            return Error::from_string_literal("Invalid integer, valid characters are 0-9 and -");
+        }
+    }
+    // The BEP says there's no limit to integer size but let's keep it to i64 here.
+    auto ret = AK::StringUtils::convert_to_int<i64>(integer_str.string_view());
+    if (ret.has_value())
+        return ret.value();
+    else
+        return Error::from_string_literal("Invalid integer, likely out of bound");
+}
+
+ErrorOr<ByteBuffer> BDecoder::parse_byte_array(Stream& stream, u8 first_byte)
+{
+    auto array_size_str = StringBuilder();
+    u8 digit = first_byte;
+    do {
+        // TODO: limit on the size of the array
+        if (digit >= '0' && digit <= '9') {
+            array_size_str.append(digit);
+        } else {
+            return Error::from_string_literal("Invalid byte array size");
+        }
+    } while ((digit = TRY(stream.read_value<u8>())) != ':');
+    auto array_size = AK::StringUtils::convert_to_uint<u64>(array_size_str.string_view(), AK::TrimWhitespace::No).value();
+    auto buffer = TRY(ByteBuffer::create_uninitialized(array_size));
+    TRY(stream.read_until_filled(buffer));
+
+    return buffer;
+}
+
+ErrorOr<Dict> BDecoder::parse_dictionary(Stream& stream)
+{
+    auto dict = Dict();
+    auto previous_key = DeprecatedString::empty();
+    u8 next_byte;
+    while ((next_byte = TRY(stream.read_value<u8>())) != 'e') {
+        auto buffer = TRY(parse_byte_array(stream, next_byte)); // key is always expected to be a byte array (string).
+        auto key = TRY(DeprecatedString::from_utf8(buffer.bytes()));
+        if (key < previous_key)
+            warnln("Invalid dictionary: entries key must be sorted"); // but many trackers don't sort them.
+        BEncodingType const& value = TRY(parse_bencoded(stream, nullptr));
+        dict.set(key, value);
+        previous_key = key;
+    }
+    return dict;
+}
+
+ErrorOr<List> BDecoder::parse_list(Stream& stream)
+{
+    // TODO: check if items are all the same type
+    auto list = List();
+    u8 next_byte;
+    while ((next_byte = TRY(stream.read_value<u8>())) != 'e') {
```

Comment:
```cpp
u8 next_byte;
while ((next_byte = TRY(stream.read_value<u8>())) != 'e')
```
That pattern seems to appear quite often, maybe we can abstract it in a function like:
`for_each_byte_until(char, Function<...>)`